### ISSUE TITLE
Cannot invoke method getServerUrl() on null when invoking openshift.withCluster in a parallel pipeline

### ISSUE THREAD

**Issue by bernardgut** [openshift/jenkins-client-plugin#239]
Hello

This is possibly a duplicate of 206.

When you do
 
```
pipeline{
  stages{
    stage('Do parallel stuff on my cluster'){
      parallel{
        stage('build something 1'){
          agent any
          steps{
            script {
              openshift.withCluster() {
                openshift.withProject() {
                ...
                }
              }
            }
          }
        }
        stage('build something 2'){
          agent any
          steps{
            script {
              openshift.withCluster() {
                openshift.withProject() {
                ...
                }
              }
            }
          }
        }
        stage('build something 3'){
          agent any
          steps{
            script {
              openshift.withCluster() {
                openshift.withProject() {
                ...
                }
              }
            }
          }
        }
      }
    }
  }
}
```

Towards the end of any of the parallel tasks you systematically get 

> [2019-02-25T07:46:21.162Z] Successfully built 3904f288b29a
> [2019-02-25T07:46:21.162Z] Pushing image XXX XXX
> [2019-02-25T07:46:21.162Z] Push successful
> [2019-02-25T07:46:21.162Z] 
> [2019-02-25T07:46:21.162Z] 
> Cannot invoke method getServerUrl() on null object

The full stack:

> java.lang.NullPointerException: Cannot invoke method getServerUrl() on null object
> 	at org.codehaus.groovy.runtime.NullObject.invokeMethod(NullObject.java:91)
> 	at org.codehaus.groovy.runtime.callsite.PogoMetaClassSite.call(PogoMetaClassSite.java:48)
> 	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:48)
> 	at org.codehaus.groovy.runtime.callsite.NullCallSite.call(NullCallSite.java:35)
> 	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:48)
> 	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:113)
> 	at com.cloudbees.groovy.cps.sandbox.DefaultInvoker.methodCall(DefaultInvoker.java:20)
> 	at com.openshift.jenkins.plugins.OpenShiftDSL.buildCommonArgs(jar:file:/var/lib/jenkins/plugins/openshift-client/WEB-INF/lib/openshift-client.jar!/com/openshift/jenkins/plugins/OpenShiftDSL.groovy:482)
> 	at com.openshift.jenkins.plugins.OpenShiftDSL.buildCommonArgs(jar:file:/var/lib/jenkins/plugins/openshift-client/WEB-INF/lib/openshift-client.jar!/com/openshift/jenkins/plugins/OpenShiftDSL.groovy:429)
> 	at com.openshift.jenkins.plugins.OpenShiftDSL.simplePassthrough(jar:file:/var/lib/jenkins/plugins/openshift-client/WEB-INF/lib/openshift-client.jar!/com/openshift/jenkins/plugins/OpenShiftDSL.groovy:979)
> 	at com.openshift.jenkins.plugins.OpenShiftDSL.tag(jar:file:/var/lib/jenkins/plugins/openshift-client/WEB-INF/lib/openshift-client.jar!/com/openshift/jenkins/plugins/OpenShiftDSL.groovy:992)
> 	at WorkflowScript.run(WorkflowScript:107)
> 	at com.openshift.jenkins.plugins.OpenShiftDSL.withProject(jar:file:/var/lib/jenkins/plugins/openshift-client/WEB-INF/lib/openshift-client.jar!/com/openshift/jenkins/plugins/OpenShiftDSL.groovy:395)
> 	at com.openshift.jenkins.plugins.OpenShiftDSL$Context.run(jar:file:/var/lib/jenkins/plugins/openshift-client/WEB-INF/lib/openshift-client.jar!/com/openshift/jenkins/plugins/OpenShiftDSL.groovy:128)
> 	at com.openshift.jenkins.plugins.OpenShiftDSL.withProject(jar:file:/var/lib/jenkins/plugins/openshift-client/WEB-INF/lib/openshift-client.jar!/com/openshift/jenkins/plugins/OpenShiftDSL.groovy:394)
> 	at WorkflowScript.run(WorkflowScript:90)
> 	at com.openshift.jenkins.plugins.OpenShiftDSL.withCluster(jar:file:/var/lib/jenkins/plugins/openshift-client/WEB-INF/lib/openshift-client.jar!/com/openshift/jenkins/plugins/OpenShiftDSL.groovy:366)
> 	at com.openshift.jenkins.plugins.OpenShiftDSL$Context.run(jar:file:/var/lib/jenkins/plugins/openshift-client/WEB-INF/lib/openshift-client.jar!/com/openshift/jenkins/plugins/OpenShiftDSL.groovy:128)
> 	at com.openshift.jenkins.plugins.OpenShiftDSL.withCluster(jar:file:/var/lib/jenkins/plugins/openshift-client/WEB-INF/lib/openshift-client.jar!/com/openshift/jenkins/plugins/OpenShiftDSL.groovy:365)
> 	at com.openshift.jenkins.plugins.OpenShiftDSL.withCluster(jar:file:/var/lib/jenkins/plugins/openshift-client/WEB-INF/lib/openshift-client.jar!/com/openshift/jenkins/plugins/OpenShiftDSL.groovy:383)
> 	at com.openshift.jenkins.plugins.OpenShiftDSL.node(jar:file:/var/lib/jenkins/plugins/openshift-client/WEB-INF/lib/openshift-client.jar!/com/openshift/jenkins/plugins/OpenShiftDSL.groovy:1791)
> 	at com.openshift.jenkins.plugins.OpenShiftDSL.withCluster(jar:file:/var/lib/jenkins/plugins/openshift-client/WEB-INF/lib/openshift-client.jar!/com/openshift/jenkins/plugins/OpenShiftDSL.groovy:370)
> 	at WorkflowScript.run(WorkflowScript:89)
> 	at ___cps.transform___(Native Method)
> 	at com.cloudbees.groovy.cps.impl.ContinuationGroup.methodCall(ContinuationGroup.java:57)
> 	at com.cloudbees.groovy.cps.impl.FunctionCallBlock$ContinuationImpl.dispatchOrArg(FunctionCallBlock.java:109)
> 	at com.cloudbees.groovy.cps.impl.FunctionCallBlock$ContinuationImpl.fixName(FunctionCallBlock.java:77)
> 	at sun.reflect.GeneratedMethodAccessor117.invoke(Unknown Source)
> 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> 	at java.lang.reflect.Method.invoke(Method.java:498)
> 	at com.cloudbees.groovy.cps.impl.ContinuationPtr$ContinuationImpl.receive(ContinuationPtr.java:72)
> 	at com.cloudbees.groovy.cps.impl.ConstantBlock.eval(ConstantBlock.java:21)
> 	at com.cloudbees.groovy.cps.Next.step(Next.java:83)
> 	at com.cloudbees.groovy.cps.Continuable$1.call(Continuable.java:174)
> 	at com.cloudbees.groovy.cps.Continuable$1.call(Continuable.java:163)
> 	at org.codehaus.groovy.runtime.GroovyCategorySupport$ThreadCategoryInfo.use(GroovyCategorySupport.java:129)
> 	at org.codehaus.groovy.runtime.GroovyCategorySupport.use(GroovyCategorySupport.java:268)
> 	at com.cloudbees.groovy.cps.Continuable.run0(Continuable.java:163)
> 	at org.jenkinsci.plugins.workflow.cps.SandboxContinuable.access$101(SandboxContinuable.java:34)
> 	at org.jenkinsci.plugins.workflow.cps.SandboxContinuable.lambda$run0$0(SandboxContinuable.java:59)
> 	at org.jenkinsci.plugins.scriptsecurity.sandbox.groovy.GroovySandbox.runInSandbox(GroovySandbox.java:121)
> 	at org.jenkinsci.plugins.workflow.cps.SandboxContinuable.run0(SandboxContinuable.java:58)
> 	at org.jenkinsci.plugins.workflow.cps.CpsThread.runNextChunk(CpsThread.java:182)
> 	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup.run(CpsThreadGroup.java:332)
> 	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup.access$200(CpsThreadGroup.java:83)
> 	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup$2.call(CpsThreadGroup.java:244)
> 	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup$2.call(CpsThreadGroup.java:232)
> 	at org.jenkinsci.plugins.workflow.cps.CpsVmExecutorService$2.call(CpsVmExecutorService.java:64)
> 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
> 	at hudson.remoting.SingleLaneExecutorService$1.run(SingleLaneExecutorService.java:131)
> 	at jenkins.util.ContextResettingExecutorService$1.run(ContextResettingExecutorService.java:28)
> 	at jenkins.security.ImpersonatingExecutorService$1.run(ImpersonatingExecutorService.java:59)
> 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
> 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
> 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
> 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
> 	at java.lang.Thread.run(Thread.java:748)

There is a race condition on the openshift global variable. 

The question is the following : Is there any way I can run parallel tasks (builds) triggered from jenkins on an openshift cluster?

thanks
- Bernard

**Comment by bernardgut** [comment]
Note that I have now seen the docs on "Compatibility with parallel step". I am just not clear whether the example provided actually runs the steps in parallel or whether the lock means it will in practice run it sequentially...

Can someone clarify?

thanks
- Bernard

**Comment by gabemontero** [comment]
The example provide runs in parallel in parallel, but you have to ensure the parallel closure is within the openshift.withCluster / openshift.withProject closures, as the example shows.

If you have a form like 
```
parallel {
   openshift.withCluster() {
   }
}
```

you *HAVE* to use openshift.setLockName to serialized things
